### ISSUE TITLE
Raw_json is null when providing a stream parameter to ChatCompletion->create_async

### ISSUE THREAD

**Issue by divonelnc** [D7EAD/liboai#29]
### Describe the bug

When adding a stream callback to the simple async conversation example, the `raw_json` is null, and an exception happens upon calling `GetLastResponse()`

### To Reproduce

1. Create a new project
2. Use the example code for Chat completion
3. Add a stream function to the call
4. Attempt to output the last response

### Code snippets

```C++
#include <iostream>

#include "include/liboai.h"

using namespace liboai;

bool streamResponse(std::string output, intptr_t context)
{
	std::cout << output << std::endl;
	return true;
}

int main()
{
    OpenAI oai;
    Conversation convo;

    convo.AddUserData("Hello, how are you?");

    if (oai.auth.SetKeyEnv("OPENAI_API_KEY")) {
        
        std::optional<std::function<bool(std::string, intptr_t)>> stream = streamResponse;

        FutureResponse fut = oai.ChatCompletion->create_async(
	        "gpt-3.5-turbo",
	        convo,
	        std::nullopt,
	        std::nullopt,
	        std::nullopt,
	        stream
        );
        
        fut.wait();
        Response response = fut.get();
        
        std::cout << "Raw Response: " << response.raw_json << std::endl;
        
        convo.Update(response);
        
        std::cout << convo.GetLastResponse();
    }
}
```

### OS

Windows 10

### Library version

liboai 3.0.2

**Comment by divonelnc** [comment]
Hello, any clues to what might be the issue?

**Comment by D7EAD** [comment]
> Hello, any clues to what might be the issue?

At this moment, I am unsure; however, I have yet to take a deep dive into it due to personal constraints. 

However, knowing how streamed data works with OpenAI and SSE, I can certainly see why this would cause issues--especially with the `Conversation` object. As of right now, there is no library-included way to parse a response from a streamed request to ChatCompletion and load it into a `Conversation` object like you would normally. This is due to the fact that SSEs and streamed data do not adhere to the JSON structure that `Conversation` expects--the structure found in synchronous calls.

`raw_json` in `Conversation` being NULL is by design, as the data that streamed calls send back is not the formatted JSON that `Conversation` expects, so it assumes it is not JSON, leaving `raw_json` NULL. 

I am not sure why an exception is being thrown. In my brief testing of this issue, even supplying a lambda to print streamed tokens from a response failed in a strange manner.

I'll continue to investigate this as much as I can. Feel free to investigate yourself as well.

**Comment by divonelnc** [comment]
I don't have much knowledge of C++, as this is literally my first C++ project. After a back-and-forth of debugging with GPT-4, it wrote itself the following code that properly streams an answer to the console.

I ran it myself, so it should work as-is in a new project. You just need to set the "YOUR-API-KEY" value.

Hope that can help:

```
#include <iostream>
#include <string>
#include <sstream>
#include <curl/curl.h>
#include <nlohmann/json.hpp>

using json = nlohmann::json;

static size_t WriteCallback(void* contents, size_t size, size_t nmemb, void* userp) {
	const std::string raw_response((char*)contents, size * nmemb);
	std::stringstream raw_response_stream(raw_response);
	std::string line;

	while (std::getline(raw_response_stream, line)) {
		if (line.substr(0, 5) == "data:") {
			std::string json_data = line.substr(6);

			if (json_data != "[DONE]") {
				try {
					json response = json::parse(json_data);

					if (response.contains("error")) {
						std::cout << "Error: " << response["error"]["message"].get<std::string>() << std::endl;
					}
					else {
						if (response.contains("choices") && response["choices"].is_array() && response["choices"].size() > 0) {
							auto message_delta = response["choices"][0]["delta"];
							if (message_delta.contains("content")) {
								std::cout << message_delta["content"].get<std::string>();
							}
						}
					}
				}
				catch (json::parse_error& e) {
					std::cout << "Parse error: " << e.what() << std::endl;
				}
			}
		}
	}

	return size * nmemb;
}

std::vector<std::string> split_response_string(const std::string& response) {
	std::vector<std::string> response_lines;
	std::istringstream stream(response);
	std::string line;

	while (std::getline(stream, line)) {
		if (line.substr(0, 5) == "data:") {
			if (line.substr(6) != "[DONE]") {
				response_lines.push_back(line.substr(6));
			}
		}
	}

	return response_lines;
}

std::string send_request(json payload) {
	std::string api_key = "YOUR-API-KEY";
	std::string read_buffer;
	CURL* curl;
	CURLcode res;

	curl_global_init(CURL_GLOBAL_DEFAULT);
	curl = curl_easy_init();

	if (curl) {
		struct curl_slist* headers = NULL;
		headers = curl_slist_append(headers, "Content-Type: application/json");
		headers = curl_slist_append(headers, ("Authorization: Bearer " + api_key).c_str());

		// Convert JSON payload to a string.
		std::string payload_string = payload.dump();

		curl_easy_setopt(curl, CURLOPT_URL, "https://api.openai.com/v1/chat/completions");
		curl_easy_setopt(curl, CURLOPT_POST, 1L);
		curl_easy_setopt(curl, CURLOPT_POSTFIELDS, payload_string.c_str());
		curl_easy_setopt(curl, CURLOPT_POSTFIELDSIZE, payload_string.length());
		curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, WriteCallback);
		curl_easy_setopt(curl, CURLOPT_WRITEDATA, &read_buffer);
		curl_easy_setopt(curl, CURLOPT_HTTPHEADER, headers);

		res = curl_easy_perform(curl);
		if (res != CURLE_OK) {
			fprintf(stderr, "curl_easy_perform() failed: %s\n", curl_easy_strerror(res));
		}

		curl_easy_cleanup(curl);
		curl_slist_free_all(headers);
	}

	return read_buffer;
}


int main() {
	json request_payload;
	request_payload["messages"] = json::array({ json::object({
		{ "role", "user" },
		{ "content", "Write a short poem about the capital of France." }
	}) });
	request_payload["model"] = "gpt-4";
	request_payload["stream"] = true;

	send_request(request_payload);

	return 0;
}
```

**Comment by D7EAD** [comment]
Hi.

After updating my installation of cURL to v8.1.2 via VCPKG, it appears that crashes have been resolved. However, I still am unable to pass a stream callback to `ChatCompletion`'s `create_async` method and have any data be printed. I will continue to investigate this and post updates here.

Feel free to also update your cURL package and experiment.

**Comment by divonelnc** [comment]
If that helps: the code I posted works, and I have been using it for some time now.

**Comment by D7EAD** [comment]
Hi.

The issues with stream callbacks have been fixed in #32 and can be found on branch `v3.1.0-dev`. In regard to SSE parsing for `Conversation`, this will likely be added later.
